### PR TITLE
Add client_cert_token_timeout to address GH-9465

### PR DESCRIPTION
**Pull request by bnfbiz**
This PR adds the parameter `client_cert_token_timeout` which sets how long the expire time on the token that is created when using a `client_cert_path` certificate.

Closes https://github.com/hashicorp/packer/issues/9465

### CODE REVIEW

**Review by sylviamoss — CHANGES_REQUESTED** [review]
It seems that your code has some generation problems. Have you updated the hcl2spec and docs manually? They should be updated automatically by running `make generate`. 

Also, we can't merge this until you get the CLA signed; let me know if you have issues with it.

**Comment by bnfbiz** [comment]
@sylviamoss I am looking at the check-generate failure.  From what I tell they failures are due to the additional attribute that the pull request added. 

Error in provisioner/azure-dtlartifact/provisioner.hcl2spec.go
All entries in the structure were moved over one character due to the lenght of the new parameter name.  Also the addition of the new parameter (second piece of this same diff).
```text
diff --git a/provisioner/azure-dtlartifact/provisioner.hcl2spec.go b/provisioner/azure-dtlartifact/provisioner.hcl2spec.go
index 95601f797..dc914de5c 100644
--- a/provisioner/azure-dtlartifact/provisioner.hcl2spec.go
+++ b/provisioner/azure-dtlartifact/provisioner.hcl2spec.go
@@ -37,30 +37,31 @@ func (*FlatArtifactParameter) HCL2Spec() map[string]hcldec.Spec {
 // FlatConfig is an auto-generated flat version of Config.
 // Where the contents of a field with a `mapstructure:,squash` tag are bubbled up.
 type FlatConfig struct {
-	PackerBuildName        *string                `mapstructure:"packer_build_name" cty:"packer_build_name" hcl:"packer_build_name"`
-	PackerBuilderType      *string                `mapstructure:"packer_builder_type" cty:"packer_builder_type" hcl:"packer_builder_type"`
-	PackerCoreVersion      *string                `mapstructure:"packer_core_version" cty:"packer_core_version" hcl:"packer_core_version"`
-	PackerDebug            *bool                  `mapstructure:"packer_debug" cty:"packer_debug" hcl:"packer_debug"`
-	PackerForce            *bool                  `mapstructure:"packer_force" cty:"packer_force" hcl:"packer_force"`
-	PackerOnError          *string                `mapstructure:"packer_on_error" cty:"packer_on_error" hcl:"packer_on_error"`
-	PackerUserVars         map[string]string      `mapstructure:"packer_user_variables" cty:"packer_user_variables" hcl:"packer_user_variables"`
-	PackerSensitiveVars    []string               `mapstructure:"packer_sensitive_variables" cty:"packer_sensitive_variables" hcl:"packer_sensitive_variables"`
-	CloudEnvironmentName   *string                `mapstructure:"cloud_environment_name" required:"false" cty:"cloud_environment_name" hcl:"cloud_environment_name"`
-	ClientID               *string                `mapstructure:"client_id" cty:"client_id" hcl:"client_id"`
-	ClientSecret           *string                `mapstructure:"client_secret" cty:"client_secret" hcl:"client_secret"`
-	ClientCertPath         *string                `mapstructure:"client_cert_path" cty:"client_cert_path" hcl:"client_cert_path"`
-	ClientJWT              *string                `mapstructure:"client_jwt" cty:"client_jwt" hcl:"client_jwt"`
-	ObjectID               *string                `mapstructure:"object_id" cty:"object_id" hcl:"object_id"`
-	TenantID               *string                `mapstructure:"tenant_id" required:"false" cty:"tenant_id" hcl:"tenant_id"`
-	SubscriptionID         *string                `mapstructure:"subscription_id" cty:"subscription_id" hcl:"subscription_id"`
-	UseAzureCLIAuth        *bool                  `mapstructure:"use_azure_cli_auth" required:"false" cty:"use_azure_cli_auth" hcl:"use_azure_cli_auth"`
-	DtlArtifacts           []FlatDtlArtifact      `mapstructure:"dtl_artifacts" cty:"dtl_artifacts" hcl:"dtl_artifacts"`
-	LabName                *string                `mapstructure:"lab_name" cty:"lab_name" hcl:"lab_name"`
-	ResourceGroupName      *string                `mapstructure:"lab_resource_group_name" cty:"lab_resource_group_name" hcl:"lab_resource_group_name"`
-	VMName                 *string                `mapstructure:"vm_name" cty:"vm_name" hcl:"vm_name"`
-	PollingDurationTimeout *string                `mapstructure:"polling_duration_timeout" required:"false" cty:"polling_duration_timeout" hcl:"polling_duration_timeout"`
-	AzureTags              map[string]*string     `mapstructure:"azure_tags" cty:"azure_tags" hcl:"azure_tags"`
-	Json                   map[string]interface{} `cty:"json" hcl:"json"`
+	PackerBuildName         *string                `mapstructure:"packer_build_name" cty:"packer_build_name" hcl:"packer_build_name"`
+	PackerBuilderType       *string                `mapstructure:"packer_builder_type" cty:"packer_builder_type" hcl:"packer_builder_type"`
+	PackerCoreVersion       *string                `mapstructure:"packer_core_version" cty:"packer_core_version" hcl:"packer_core_version"`
+	PackerDebug             *bool                  `mapstructure:"packer_debug" cty:"packer_debug" hcl:"packer_debug"`
+	PackerForce             *bool                  `mapstructure:"packer_force" cty:"packer_force" hcl:"packer_force"`
+	PackerOnError           *string                `mapstructure:"packer_on_error" cty:"packer_on_error" hcl:"packer_on_error"`
+	PackerUserVars          map[string]string      `mapstructure:"packer_user_variables" cty:"packer_user_variables" hcl:"packer_user_variables"`
+	PackerSensitiveVars     []string               `mapstructure:"packer_sensitive_variables" cty:"packer_sensitive_variables" hcl:"packer_sensitive_variables"`
+	CloudEnvironmentName    *string                `mapstructure:"cloud_environment_name" required:"false" cty:"cloud_environment_name" hcl:"cloud_environment_name"`
+	ClientID                *string                `mapstructure:"client_id" cty:"client_id" hcl:"client_id"`
+	ClientSecret            *string                `mapstructure:"client_secret" cty:"client_secret" hcl:"client_secret"`
+	ClientCertPath          *string                `mapstructure:"client_cert_path" cty:"client_cert_path" hcl:"client_cert_path"`
+	ClientCertExpireTimeout *string                `mapstructure:"client_cert_token_timeout" required:"false" cty:"client_cert_token_timeout" hcl:"client_cert_token_timeout"`
+	ClientJWT               *string                `mapstructure:"client_jwt" cty:"client_jwt" hcl:"client_jwt"`
+	ObjectID                *string                `mapstructure:"object_id" cty:"object_id" hcl:"object_id"`
+	TenantID                *string                `mapstructure:"tenant_id" required:"false" cty:"tenant_id" hcl:"tenant_id"`
+	SubscriptionID          *string                `mapstructure:"subscription_id" cty:"subscription_id" hcl:"subscription_id"`
+	UseAzureCLIAuth         *bool                  `mapstructure:"use_azure_cli_auth" required:"false" cty:"use_azure_cli_auth" hcl:"use_azure_cli_auth"`
+	DtlArtifacts            []FlatDtlArtifact      `mapstructure:"dtl_artifacts" cty:"dtl_artifacts" hcl:"dtl_artifacts"`
+	LabName                 *string                `mapstructure:"lab_name" cty:"lab_name" hcl:"lab_name"`
+	ResourceGroupName       *string                `mapstructure:"lab_resource_group_name" cty:"lab_resource_group_name" hcl:"lab_resource_group_name"`
+	VMName                  *string                `mapstructure:"vm_name" cty:"vm_name" hcl:"vm_name"`
+	PollingDurationTimeout  *string                `mapstructure:"polling_duration_timeout" required:"false" cty:"polling_duration_timeout" hcl:"polling_duration_timeout"`
+	AzureTags               map[string]*string     `mapstructure:"azure_tags" cty:"azure_tags" hcl:"azure_tags"`
+	Json                    map[string]interface{} `cty:"json" hcl:"json"`
 }
 
 // FlatMapstructure returns a new FlatConfig.
@@ -87,6 +88,7 @@ func (*FlatConfig) HCL2Spec() map[string]hcldec.Spec {
 		"client_id":                  &hcldec.AttrSpec{Name: "client_id", Type: cty.String, Required: false},
 		"client_secret":              &hcldec.AttrSpec{Name: "client_secret", Type: cty.String, Required: false},
 		"client_cert_path":           &hcldec.AttrSpec{Name: "client_cert_path", Type: cty.String, Required: false},
+		"client_cert_token_timeout":  &hcldec.AttrSpec{Name: "client_cert_token_timeout", Type: cty.String, Required: false},
 		"client_jwt":                 &hcldec.AttrSpec{Name: "client_jwt", Type: cty.String, Required: false},
 		"object_id":                  &hcldec.AttrSpec{Name: "object_id", Type: cty.String, Required: false},
 		"tenant_id":                  &hcldec.AttrSpec{Name: "tenant_id", Type: cty.String, Required: false},
```

Error in website/content/partials/builder/azure/common/client/Config-not-required.mdx
The difference again is due to the documentation for the new parameter being added to the file (when it was generated).
```text
diff --git a/website/content/partials/builder/azure/common/client/Config-not-required.mdx b/website/content/partials/builder/azure/common/client/Config-not-required.mdx
index 46d553c81..eca7e1ef8 100644
--- a/website/content/partials/builder/azure/common/client/Config-not-required.mdx
+++ b/website/content/partials/builder/azure/common/client/Config-not-required.mdx
@@ -12,6 +12,8 @@
 - `client_cert
_path` (string) - The path to a pem-encoded certificate that will be used to authenticate
   as the specified AAD SP.
 
+- `client_cert_token_timeout` (duration string | ex: "1h5m2s") - The timeout for the JWT Token when using a client certificate
+
 - `client_jwt` (string) - A JWT bearer token for client auth (RFC 7523, Sec. 2.2) that will be used
   to authenticate the AAD SP. Provides more control over token the expiration
   when using certificate authentication than when using `client_cert_path`.
Found diffs in go generated code.
You can use the command: `make generate` to reformat code.
make: *** [Makefile:131: generate-check] Error 1
```

Is there something that I should be doing differently or what am I missing?

**Comment by sylviamoss** [comment]
You had to run `make generate` to propper generate the docs and the hcl2spec for the azure provisioner that uses the same config you've changed. I just pushed to your branch the fix for that. 

You need to sign the CLA so that we can merge this once is approved.

**Comment by bnfbiz** [comment]
Sorry I missed the point you were saying to add the files to the commit.  I am working on the CLA (legal has to review it).

**Comment by bnfbiz** [comment]
@azr I got part of my approvals still need one more  to be allowed to sign the form.  Meanwhile I was thinking about unit tests.   However, I am not well versed in that part and was wondering if adding this to `builder/azure/common/client/config_test.go` maybe somewhere around line 98 would be a proper unit test:

```go
{
			name: "client_cert_token_timeout used when using JWT should error",
			config: Config{
				ClientID:                "ok",
				ClientJWT:               "ok",
				ClientCertExpireTimeout: 3600,
			},
			wantErr: true,
		},
```

What I am trying to test is that it would be an error to have the certificate expire timeout parameter set with a user provided JWT.

Thanks for any guidance.

**Comment by bnfbiz** [comment]
@sylviamoss  and @paulmey what is the process to get this approved or if there are changes needed?

Thanks

**Comment by bnfbiz** [comment]
> It seems that your code has some generation problems. Have you updated the hcl2spec and docs manually? They should be updated automatically by running `make generate`.
> 
> Also, we can't merge this until you get the CLA signed; let me know if you have issues with it.
both of these issues are resolved

**Comment by sylviamoss** [comment]
hey @bnfbiz, thank you for pinging and sorry for the delay. I'll try to review it today. 
It would be ideal to have the code owner approval as well (@paulmey)

**Review by sylviamoss — COMMENTED** [review]
This is looking nice! I haven't tested though, I need to recover my azure credentials.

Regarding the test you purposed. I think is a good idea but the case is not getting the correct error. 
The correct case would be: 
```go
{
	name: "client_cert_token_timeout used when using JWT should error",
	config: Config{
		SubscriptionID:          "ok",
		ClientID:                "ok",
		ClientJWT:                getJWT(10*time.Minute, true),
		ClientCertExpireTimeout: 3600,
	},
	wantErr: true,
},
```

Right now this will fail the test. You need to implement the validation inside this condition https://github.com/hashicorp/packer/blob/master/builder/azure/common/client/config.go#L175.

**Comment by sylviamoss** [comment]
I see that the minimum exp time a provided JWT must have is 5min. I think we should check that is `client_cert_token_timeout` is bigger than 5min as well.

File: builder/azure/common/client/config.go
Comment (bnfbiz) [reply builder/azure/common/client/config.go]: I updated the test for the client_cert_token_timeout to be atleast 5 minutes.  I also rebased off master (hope that was OK).  I am running a local azure build that will time out in 10860 minutes (3hrs 1 min) with `client_cert_token_timeout = "4h"`

I also ran client tests
```shell
$ make test TEST=./builder/azure/common/client
==> Checking that only certain files are executable...
Check passed.
ok      github.com/hashicorp/packer/builder/azure/common/client 1.020s
```

**Comment by bnfbiz** [comment]
> I see that the minimum exp time a provided JWT must have is 5min. I think we should check that is `client_cert_token_timeout` is bigger than 5min as well.

I pushed a test for this.

File: builder/azure/common/client/config_test.go
Comment (sylviamoss) [inline builder/azure/common/client/config_test.go:99]: Nice!

**Review by sylviamoss — APPROVED** [review]
This looks good! I pushed the fix for the linter and generated code.

**Pull request merged**
